pkg/observable/channel/replay.go

```diff
@@ -0,0 +1,209 @@
+	ctx context.Context,
+	replayBufferSize int,
+	srcObsvbl observable.Observable[V],
+) observable.ReplayObservable[V] {
+	// Assert that the source observable implements the internals required to
+	// embed and wrap it.
+	internals := srcObsvbl.(observableInternals[V])
+
+	replayObsvbl := &replayObservable[V]{
+		observableInternals: internals,
+		replayBufferSize:    replayBufferSize,
+		replayBuffer:        make([]V, 0, replayBufferSize),
+	}
+
+	srcObserver := srcObsvbl.Subscribe(ctx)
+	go replayObsvbl.goBufferReplayNotifications(srcObserver)
+
+	return replayObsvbl
+}
+
+// Last synchronously returns the last n values from the replay buffer. It blocks
+// until at least 1 notification has been accumulated, then waits replayPartialBufferTimeout
+// duration before returning all notifications accumulated notifications by that time.
+// If the replay buffer contains at least n notifications, this function will only
+// block as long as it takes to accumulate and return them.
+// If n is greater than the replay buffer size, the entire replay buffer is returned.
+func (ro *replayObservable[V]) Last(ctx context.Context, n int) []V {
+	// Use a temporary observer to accumulate replay values.
+	// Subscribe will always start with the replay buffer, so we can safely
+	// leverage it here for syncrhonization (i.e. blocking until at least 1
+	// notification has been accumulated). This also eliminates the need for
+	// locking and/or copying the replay buffer.
+	tempObserver := ro.Subscribe(ctx)
+	defer tempObserver.Unsubscribe()
+
+	// If n is greater than the replay buffer size, return the entire replay buffer.
+	if n > ro.replayBufferSize {
+		n = ro.replayBufferSize
+		log.Printf(
+			"WARN: requested replay buffer size %d is greater than replay buffer capacity %d; returning entire replay buffer",
+			n, cap(ro.replayBuffer),
+		)
+	}
+
+	// Accumulate replay values in a new slice to avoid (read) locking replayBufferMu.
+	var values []V
+	// accumulateReplayValues works concurrently and returns a context and cancellation
+	// function for signaling completion.
+	valuesAccCtx, valuesAccCancel := accumulateReplayValues(tempObserver, n, &values)
+
+	// Wait for N values to be accumulated or timeout. When timing out, if we
+	// have at least 1 value, we can return it. Otherwise, we need to wait for
+	// the next value to be published (i.e. continue the loop).
+	for {
+		select {
+		case <-valuesAccCtx.Done():
+			return values
+		case <-time.After(replayPartialBufferTimeout):
+			if len(values) > 1 {
+				valuesAccCancel()
+				return values
+			}
+		}
+	}
+}
+
+// Subscribe returns an observer which is notified when the publishCh channel
+// receives a value.
+func (ro *replayObservable[V]) Subscribe(ctx context.Context) observable.Observer[V] {
+	ro.replayBufferMu.RLock()
+	defer ro.replayBufferMu.RUnlock()
+
+	observer := NewObserver[V](ctx, ro.onUnsubscribe)
+
+	// ToReplayObservable all buffered replayBuffer to the observer channel buffer before
```

Comment:
Is this comment out of place or outdated? Can't understand what `ToReplayObservable` is intended for here.

Reply:
Nice catch! I think this was my IDE getting carried away when renaming `Replay` to `ToReplayObservable`.